MapsetVerifier.Checks/Utils/TaikoUtils.cs

```diff
@@ -0,0 +1,199 @@
+            var barlineGap = timing.msPerBeat * timing.Meter;
+
+            var offsetFromNextImplicitBarline = ((time - timing.Offset) % barlineGap) - barlineGap;
+            var nextPotentialRedLine = beatmap.GetTimingLine<UninheritedLine>(time + offsetFromNextImplicitBarline);
+            var offsetFromNextPotentialRedline = time - nextPotentialRedLine.Offset;
+
+            return TakeLowerAbsValue(offsetFromNextImplicitBarline, offsetFromNextPotentialRedline);
+        }
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the nearest barline to the given time.
+        /// </summary>
+        public static double GetOffsetFromNearestBarlineMs(Beatmap beatmap, double time)
+        {
+            return TakeLowerAbsValue(GetOffsetFromPrevBarlineMs(beatmap, time), GetOffsetFromNextBarlineMs(beatmap, time));
+        }
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the previous barline to the hit object's head (start time).
+        /// </summary>
+        public static double GetHeadOffsetFromPrevBarlineMs(this HitObject current) => GetOffsetFromPrevBarlineMs(current.beatmap, current.time);
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the next barline to the hit object's head (start time).
+        /// </summary>
+        public static double GetHeadOffsetFromNextBarlineMs(this HitObject current) => GetOffsetFromNextBarlineMs(current.beatmap, current.time);
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the nearest barline to the hit object's head (start time).
+        /// </summary>
+        public static double GetHeadOffsetFromNearestBarlineMs(this HitObject current) => GetOffsetFromNearestBarlineMs(current.beatmap, current.time);
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the previous barline to the hit object's tail (end time).
+        /// </summary>
+        public static double GetTailOffsetFromPrevBarlineMs(this HitObject current) => GetOffsetFromPrevBarlineMs(current.beatmap, current.GetEndTime());
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the next barline to the hit object's tail (end time).
+        /// </summary>
+        public static double GetTailOffsetFromNextBarlineMs(this HitObject current) => GetOffsetFromNextBarlineMs(current.beatmap, current.GetEndTime());
+
+        /// <summary>
+        ///     Returns the offset in milliseconds from the nearest barline to the hit object's tail (end time).
+        /// </summary>
+        public static double GetTailOffsetFromNearestBarlineMs(this HitObject current) => GetOffsetFromNearestBarlineMs(current.beatmap, current.GetEndTime());
+
+        /// <summary>
+        ///     Returns whether the beatmap set contains a beatmap with Easy difficulty (Kantan).
+        /// </summary>
+        public static bool IsBottomDiffKantan(this BeatmapSet beatmapSet)
```

Comment:
It can just be a function in the PatternLengthCheck class‎ instead, as you don't seem to use it anywhere else?